File: apps/android/src/main/kotlin/space/bitos/app/ui/feed/MemePublishJobStore.kt
package space.bitos.app.ui.feed

import android.content.Context
import org.json.JSONArray
import org.json.JSONObject
import java.io.File

/**
 * Durable publish-job ledger (prototype `#/queue`): meme publish attempts
 * persist inputs + stage so a crash never loses the run. Small JSON
 * ledger + one media file per job (deleted on done/discard); bounded.
 */
data class MemePublishJob(
    val id: Int,
    val createdAtMs: Long,
    val updatedAtMs: Long,
    val mode: String, // image | gif | video
    val caption: String,
    val altText: String,
    val contentWarningReason: String?,
    val extraTagsJson: String,
    val remixEventId: String,
    val remixAuthor: String,
    val mime: String,
    val width: Int,
    val height: Int,
    val durationMs: Long,
    val coverThumbUrl: String?,
    val stage: Int, // 0…7 (render…confirm)
    val mediaUrl: String? = null,
    val mediaSha256: String? = null,
    val eventId: String? = null,
    val status: String, // active | failed | done | discarded
    val lastError: String? = null,
    val mediaFile: String,
) {
    val isTerminal: Boolean get() = status == "done" || status == "discarded"
    /** Safe to re-send only before anything signed (an event id may be live). */
    val retryAllowed: Boolean get() = eventId == null && status != "done"
}

class MemePublishJobStore(context: Context) {
    private val dir = File(context.filesDir, "meme-publish-jobs").apply { runCatching { mkdirs() } }
    private val ledger = File(dir, "jobs.json")
    private val cap = 20

    fun list(): List<MemePublishJob> = runCatching {
        val array = JSONArray(ledger.readText())
        List(array.length()) { i -> decode(array.getJSONObject(i)) }
    }.getOrDefault(emptyList())

    fun recoverable(): List<MemePublishJob> = list().filterNot { it.isTerminal }

    fun job(id: Int): MemePublishJob? = list().firstOrNull { it.id == id }

    fun begin(
        mode: String, caption: String, altText: String, cw: String?,
        extraTagsJson: String, remixEventId: String, remixAuthor: String,
        bytes: ByteArray, mime: String, width: Int, height: Int,
        durationMs: Long, coverThumbUrl: String?, nowMs: Long,
    ): Int {
        val id = ((nowMs / 1000) % 10_000).toInt()
        val file = "job-$id-$nowMs.bin"
        runCatching { File(dir, file).writeBytes(bytes) }
        val job = MemePublishJob(
            id, nowMs, nowMs, mode, caption, altText, cw, extraTagsJson,
            remixEventId, remixAuthor, mime, width, height, durationMs,
            coverThumbUrl, 0, status = "active", mediaFile = file,
        )
        persist((listOf(job) + list().filterNot { it.id == id }).take(cap))
        return id
    }

    fun update(
        id: Int, stage: Int? = null, mediaUrl: String? = null,
        sha256: String? = null, eventId: String? = null,
        status: String? = null, error: String? = null, nowMs: Long = System.currentTimeMillis(),
    ) {
        val jobs = list().toMutableList()
        val index = jobs.indexOfFirst { it.id == id }
        if (index < 0) return
        var job = jobs[index]
        stage?.let { job = job.copy(stage = it) }
        mediaUrl?.let { job = job.copy(mediaUrl = it) }
        sha256?.let { job = job.copy(mediaSha256 = it) }
        eventId?.let { job = job.copy(eventId = it) }
        status?.let { job = job.copy(status = it, lastError = if (it == "active") null else job.lastError) }
        error?.let { job = job.copy(lastError = it) }
        jobs[index] = job.copy(updatedAtMs = nowMs)
        persist(jobs)
    }

    fun finish(id: Int) {
        val jobs = list().toMutableList()
        val index = jobs.indexOfFirst { it.id == id }
        if (index < 0) return
        runCatching { File(dir, jobs[index].mediaFile).delete() }
        jobs[index] = jobs[index].copy(status = "done", stage = 8)
        persist(jobs)
    }

    fun discard(id: Int) {
        val jobs = list()
        jobs.firstOrNull { it.id == id }?.let {
            runCatching { File(dir, it.mediaFile).delete() }
        }
        persist(jobs.filterNot { it.id == id })
    }

    fun loadBytes(job: MemePublishJob): ByteArray? =
        runCatching { File(dir, job.mediaFile).readBytes() }.getOrNull()

    /** Stored bytes still hash to the recorded digest? null = media missing. */
    fun integrityOK(job: MemePublishJob): Boolean? {
        val expected = job.mediaSha256 ?: return null
        val bytes = loadBytes(job) ?: return null
        val digest = space.bitos.core.nostr.Sha256EventHasher.sha256(bytes)
            .joinToString("") { String.format("%02x", it.toInt() and 0xff) }
        return digest == expected
    }

    private fun persist(jobs: List<MemePublishJob>) {
        runCatching {
            ledger.writeText(JSONArray().apply { jobs.forEach { put(encode(it)) } }.toString())
        }
    }

    private fun encode(j: MemePublishJob) = JSONObject().apply {
        put("v", 1); put("id", j.id); put("createdAtMs", j.createdAtMs); put("updatedAtMs", j.updatedAtMs)
        put("mode", j.mode); put("caption", j.caption); put("altText", j.altText)
        put("cw", j.contentWarningReason ?: ""); put("extraTagsJson", j.extraTagsJson)
        put("remixEventId", j.remixEventId); put("remixAuthor", j.remixAuthor)
        put("mime", j.mime); put("width", j.width); put("height", j.height); put("durationMs", j.durationMs)
        put("coverThumbUrl", j.coverThumbUrl ?: ""); put("stage", j.stage)
        put("mediaUrl", j.mediaUrl ?: ""); put("sha256", j.mediaSha256 ?: ""); put("eventId", j.eventId ?: "")
        put("status", j.status); put("lastError", j.lastError ?: ""); put("mediaFile", j.mediaFile)
    }

    private fun decode(o: JSONObject) = MemePublishJob(
        id = o.getInt("id"), createdAtMs = o.getLong("createdAtMs"), updatedAtMs = o.getLong("updatedAtMs"),
        mode = o.getString("mode"), caption = o.optString("caption"), altText = o.optString("altText"),
        contentWarningReason = o.optString("cw").ifBlank { null },
        extraTagsJson = o.optString("extraTagsJson"), remixEventId = o.optString("remixEventId"),
        remixAuthor = o.optString("remixAuthor"), mime = o.getString("mime"),
        width = o.optInt("width"), height = o.optInt("height"), durationMs = o.optLong("durationMs"),
        coverThumbUrl = o.optString("coverThumbUrl").ifBlank { null }, stage = o.optInt("stage"),
        mediaUrl = o.optString("mediaUrl").ifBlank { null },
        mediaSha256 = o.optString("sha256").ifBlank { null },
        eventId = o.optString("eventId").ifBlank { null },
        status = o.getString("status"), lastError = o.optString("lastError").ifBlank { null },
        mediaFile = o.getString("mediaFile"),
    )
}

File: apps/ios/BitOS/Features/Create/MemePublishJobStore.swift
import BusinessCore
import CryptoKit
import Foundation

/// Durable publish-job ledger (prototype `#/queue`): every meme publish
/// attempt persists its inputs + stage so a crash never loses the run.
/// Records are small JSON; the rendered media rides as one file per job
/// (deleted on done/discard). Bounded to [cap] records, oldest pruned.
struct MemePublishJob: Codable, Identifiable, Equatable {
    var v: Int = 1
    var id: Int
    var createdAtMs: Int64
    var updatedAtMs: Int64
    var mode: String          // image | gif | video
    var caption: String
    var altText: String
    var contentWarningReason: String?
    var extraTagsJson: String
    var remixEventId: String
    var remixAuthor: String
    var mime: String
    var width: Int
    var height: Int
    var durationMs: Int64
    var coverThumbUrl: String?
    /// Pipeline stage index 0…7 (render…confirm) at last touch.
    var stage: Int
    var mediaUrl: String?
    var mediaSha256: String?
    var eventId: String?
    /// active | failed | done | discarded
    var status: String
    var lastError: String?
    /// Job media file name inside the store directory.
    var mediaFile: String

    var isTerminal: Bool { status == "done" || status == "discarded" }
    /// Retry is safe only before anything signed: once an event id exists
    /// the note may already be live — verify instead of re-sending.
    var retryAllowed: Bool { eventId == nil && status != "done" }
}

struct MemePublishJobStore {
    let directory: URL
    private(set) var jobs: [MemePublishJob] = []
    private let cap = 20

    init(directory: URL? = nil) {
        let base = directory ?? FileManager.default
            .urls(for: .applicationSupportDirectory, in: .userDomainMask)[0]
            .appendingPathComponent("meme-publish-jobs", isDirectory: true)
        self.directory = base
        try? FileManager.default.createDirectory(at: base, withIntermediateDirectories: true)
        load()
    }

    private func ledgerURL() -> URL { directory.appendingPathComponent("jobs.json") }
    private func mediaURL(_ job: MemePublishJob) -> URL { directory.appendingPathComponent(job.mediaFile) }

    private mutating func load() {
        guard let data = try? Data(contentsOf: ledgerURL()),
              let decoded = try? JSONDecoder().decode([MemePublishJob].self, from: data) else { return }
        jobs = decoded
    }

    private func persist(_ jobs: [MemePublishJob]) {
        if let data = try? JSONEncoder().encode(jobs) {
            try? data.write(to: ledgerURL(), options: .atomic)
        }
    }

    /// Creates a job + stores the rendered bytes. Returns the job id.
    @discardableResult
    mutating func begin(
        mode: String, caption: String, altText: String, contentWarningReason: String?,
        extraTagsJson: String, remixEventId: String, remixAuthor: String,
        bytes: Data, mime: String, width: Int, height: Int, durationMs: Int64,
        coverThumbUrl: String?, nowMs: Int64
    ) -> Int {
        let id = Int((nowMs / 1000) % 10_000) // stable, human-scale chip number
        let file = "job-\(id)-\(nowMs).bin"
        try? bytes.write(to: directory.appendingPathComponent(file), options: .atomic)
        let job = MemePublishJob(
            id: id, createdAtMs: nowMs, updatedAtMs: nowMs,
            mode: mode, caption: caption, altText: altText,
            contentWarningReason: contentWarningReason, extraTagsJson: extraTagsJson,
            remixEventId: remixEventId, remixAuthor: remixAuthor,
            mime: mime, width: width, height: height, durationMs: durationMs,
            coverThumbUrl: coverThumbUrl, stage: 0,
            status: "active", mediaFile: file
        )
        jobs.removeAll { $0.id == id }
        jobs.insert(job, at: 0)
        jobs = Array(jobs.prefix(cap))
        persist(jobs)
        return id
    }

    /// Stage checkpoint update (render…confirm index); optional facts.
    /// `extraTagsJson` re-persists the merged extras once the remix lineage
    /// has joined the draft tags, so a queue retry republishes the same set.
    mutating func update(_ id: Int, stage: Int? = nil, mediaUrl: String? = nil,
                         sha256: String? = nil, eventId: String? = nil,
                         status: String? = nil, error: String? = nil,
                         extraTagsJson: String? = nil, nowMs: Int64) {
        guard let index = jobs.firstIndex(where: { $0.id == id }) else { return }
        if let stage { jobs[index].stage = stage }
        if let mediaUrl { jobs[index].mediaUrl = mediaUrl }
        if let extraTagsJson { jobs[index].extraTagsJson = extraTagsJson }
        if let sha256 { jobs[index].mediaSha256 = sha256 }
        if let eventId { jobs[index].eventId = eventId }
        if let status { jobs[index].status = status }
        if let error { jobs[index].lastError = error } else if status == "active" { jobs[index].lastError = nil }
        jobs[index].updatedAtMs = nowMs
        persist(jobs)
    }

    /// Done jobs drop their media (the note carries it now).
    mutating func finish(_ id: Int, nowMs: Int64) {
        guard let index = jobs.firstIndex(where: { $0.id == id }) else { return }
        jobs[index].status = "done"
        jobs[index].stage = 8
        jobs[index].updatedAtMs = nowMs
        try? FileManager.default.removeItem(at: mediaURL(jobs[index]))
        persist(jobs)
    }

    mutating func discard(_ id: Int) {
        guard let index = jobs.firstIndex(where: { $0.id == id }) else { return }
        try? FileManager.default.removeItem(at: mediaURL(jobs[index]))
        jobs.remove(at: index)
        persist(jobs)
    }

    func job(_ id: Int) -> MemePublishJob? { jobs.first { $0.id == id } }

    func loadBytes(_ job: MemePublishJob) -> Data? { try? Data(contentsOf: mediaURL(job)) }

    /// Verify integrity: the stored bytes still hash to the recorded digest.
    func integrityOK(_ job: MemePublishJob) -> Bool? {
        guard let expected = job.mediaSha256,
              let bytes = loadBytes(job) else { return nil }
        let digest = SHA256.hash(data: bytes).map { String(format: "%02x", $0) }.joined()
        return digest == expected
    }

    /// Non-terminal jobs (active or failed) — the recovery list.
    var recoverable: [MemePublishJob] { jobs.filter { !$0.isTerminal } }
}
